service.libraryautoupdate/service.py

```diff
@@ -0,0 +1,435 @@
+# -*- coding: cp1252 -*-
+import time
+from datetime import datetime
+import xbmc
+import xbmcgui
+import xbmcvfs
+import os
+import urllib2
+import json
+import resources.lib.utils as utils
+from resources.lib.croniter import croniter
+from resources.lib.cronclasses import CronSchedule, CustomPathFile
+
+UPGRADE_INT = 1  #to keep track of any upgrade notifications
+
+class AutoUpdater:
+    last_run = 0
+    sleep_time = 500
+    schedules = []
+    lock = False
+    
+    monitor = None
+    
+    #setup the timer amounts
+    timer_amounts = {}
+    timer_amounts['0'] = 1
+    timer_amounts['1'] = 2
+    timer_amounts['2'] = 4
+    timer_amounts['3'] = 6
+    timer_amounts['4'] = 12
+    timer_amounts['5'] = 24
+
+    def __init__(self):
+        utils.check_data_dir()  #in case this directory does not exist yet
+        self.monitor = UpdateMonitor(update_settings = self.createSchedules,after_scan = self.databaseUpdated)
+        self.readLastRun()
+        
+        #force and update on startup to create the array
+        self.createSchedules(True)
+        
+    def runProgram(self):
+        #a one-time catch for the startup delay
+        if(int(utils.getSetting("startup_delay")) != 0):
+            count = 0
+            while count < len(self.schedules):
+                if(time.time() > self.schedules[count].next_run):
+                    #we missed at least one update, fix this
+                    self.schedules[count].next_run = time.time() + int(utils.getSetting("startup_delay")) * 60
+                count = count + 1
+
+
+        #display upgrade messages if they exist
+        if(int(utils.getSetting('upgrade_notes')) < UPGRADE_INT):
+            xbmcgui.Dialog().ok(utils.getString(30000),utils.getString(30030))
+            utils.setSetting('upgrade_notes',str(UPGRADE_INT))
+            
+        
+        #program has started, check if we should show a notification
+        self.showNotify()
+
+        while(not xbmc.abortRequested):
+
+            #don't check unless new minute
+            if(time.time() > self.last_run + 60):
+                self.readLastRun()
+
+                self.evalSchedules()
+
+            xbmc.sleep(self.sleep_time)
+
+        #clean up monitor on exit
+        del self.monitor
+
+    def evalSchedules(self, manual=False):
+        if(not self.lock):
+            now = time.time()
+        
+            count = 0
+            while count < len(self.schedules):
+                cronJob = self.schedules[count]
+            
+                if(cronJob.next_run <= now):
+                    if(xbmc.Player().isPlaying() == False or utils.getSetting("run_during_playback") == "true"):
```

Comment:
maybe you can create a player instance before the loop and reuse it to avoid creating a new one all the time?